Sources/Root/SearchQueryViewController.swift

```diff
@@ -0,0 +1,194 @@
+        tableView.translatesAutoresizingMaskIntoConstraints = false
+        return tableView
+    }()
+}
+
+// MARK: - TableView DataSource
+
+extension SearchQueryViewController: UITableViewDataSource {
+    public func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return suggestions.count
+    }
+
+    public func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeue(SearchQuerySuggestionCell.self, for: indexPath)
+        let suggestion = suggestions[safe: indexPath.row]
+        cell.suggestionLabel.text = suggestion
+        return cell
+    }
+}
+
+// MARK: - TableView Delegate
+
+extension SearchQueryViewController: UITableViewDelegate {
+    public func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        let query = suggestions[safe: indexPath.row]
+        searchBar.text = query
+        tableView.deselectRow(at: indexPath, animated: true)
+        delegate?.searchViewController(self, didSelectQuery: query)
+        currentQuery = query
+        returnToSuperView()
+    }
+}
+
+// MARK: - SearchBar Delegate
+
+extension SearchQueryViewController: UISearchBarDelegate {
+    public func searchBarShouldBeginEditing(_ searchBar: UISearchBar) -> Bool {
+        // User clicked the x-button and cleared the text -> should not begin editing
+        guard !didClearText else {
+            didClearText = false
+            return false
+        }
+        // Present if needed
+        if searchBar.superview != view {
+            setup()
+            delegate?.presentSearchViewController(self)
+        }
+        isActive = true
+        return true
+    }
+
+    public func searchBarTextDidBeginEditing(_ searchBar: UISearchBar) {
+        searchBar.setShowsCancelButton(true, animated: false)
+    }
+
+    public func searchBarSearchButtonClicked(_ searchBar: UISearchBar) {
+        delegate?.searchViewController(self, didSelectQuery: searchBar.text)
+        currentQuery = searchBar.text
+        returnToSuperView()
+    }
+
+    public func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
+        if let currentQuery = currentQuery {
+            // return to previous search
+            delegate?.searchViewController(self, didSelectQuery: nil)
+            searchBar.text = currentQuery
+            suggestions(forSearchText: currentQuery)
+        } else {
+            delegate?.searchViewControllerDidCancelSearch(self)
+            currentQuery = nil
+            searchBar.text = nil
+            searchBar.setShowsCancelButton(false, animated: false)
+            suggestions(forSearchText: nil)
+        }
+        returnToSuperView()
+    }
+
+    public func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
+        // If not active, the user clicked the x-button while not editing and the search should be cancelled
+        if !isActive, searchText.isEmpty {
```

Comment:
I _think_ you might be able to use `searchBar.isDescendant(of: view)` instead of isActive, because when the searchBar is still inside the root table view cell that should return false

Reply:
Nice one